corehq/apps/smsforms/tests/test_app.py

```diff
@@ -27,7 +27,7 @@
 from corehq.messaging.scheduling.util import utcnow
 
 
-@patch_user_data_db_layer
+@patch_user_data_db_layer()
```

Comment:
What happens if the author forgets to call this decorator? Does it raise an error at the time the decorator is applied, or does it make the test appear to pass because the decorator was not applied correctly?

Reply:
Yikes, it seems like the entire thing is just skipped entirely. I'd imagine that's because the improperly-specified decorator returns something that the scanner doesn't identify as a test. I ended up using `ContextDecorator` rather than the function we were playing with, because `_create_es_user` uses `patch_user_data_db_layer` as a context manager, and it seemed required to use a class-based approach for this. Do you know of a way to handle the context-manager use case while also making the call optional?

Reply:
```py
def patch_user_data_db_layer(fn=None, *, user_schema=None):
    context = _patch_user_data_db_layer(user_schema=user_schema)
    return context(fn) if fn else context


class _patch_user_data_db_layer(ContextDecorator):
    ...
```

Reply:
Also, all `contextmanagers` are context decorators by default, so the `ContextDecorator` could be replaced with
```py
@contextmanager
def _patch_user_data_db_layer(user_schema=None):
    with (
        patch('corehq.apps.users.user_data.UserData.for_user',
              new=lambda u, d: UserData({}, None, d)),
        patch('corehq.apps.users.user_data.UserData._schema_defaults',
              new=PropertyMock(return_value=user_schema or {}))
    ):
        yield
```

Reply:
Thanks! 55d644503ac133e9b5c30c2e4ce0eb06275c18ef